### PR TITLE
feat(outdated): add --no-table option

### PR DESCRIPTION
**Pull request by aparajita**
This PR adds a `--no-table` option that prints out the info on each package vertically concatenated with no borders and a blank line between each package.

Note that the `--long` option may also be used in conjunction with `--no-table` to add the details at the end of each set of package info.

This PR also fixes a bug in the previous version that calculated cell widths without stripping ANSI characters first.

### CODE REVIEW

**Comment by aparajita** [comment]
@zkochan Once you're happy with this implementation I'll add the help.

File: packages/pnpm/src/cmd/outdated.ts [hunk 1]
```diff
@@ -88,42 +91,72 @@ export default async function (
 
   if (!outdatedPackages.length) return
 
-  let columnNames = [
-    'Package',
-    'Current',
-    'Latest'
-  ]
+  if (opts.table !== false) {
```
Comment (zkochan) [inline packages/pnpm/src/cmd/outdated.ts:104]: why not just `if (opts.table)` ?

File: packages/pnpm/src/cmd/outdated.ts [hunk 2]
```diff
@@ -88,42 +91,72 @@ export default async function (
...
+          if (details) {
+            info += `\n${details}`
+          }
+        }
+
+        return info
+      })
+      .join('\n\n') + '\n'
   }
+}
```
Comment (zkochan) [inline packages/pnpm/src/cmd/outdated.ts:158]: Let's divide these into two functions: `renderOutdatedTable` and `renderOutdatedList`

File: packages/pnpm/src/cmd/outdated.ts [hunk 3]
```diff
@@ -88,42 +91,72 @@ export default async function (
...
-    ...R.sortWith(
-      [
-        sortBySemverChange,
-        (o1, o2) => o1.packageName.localeCompare(o2.packageName),
-      ],
-      outdatedPackages.map(toOutdatedWithVersionDiff)
-    )
-      .map((outdatedPkg) => columnFns.map((fn) => fn(outdatedPkg))),
-  ], TABLE_OPTIONS)
+function sortedPackages (outdatedPackages: Array<OutdatedPackage>) {
```
Comment (zkochan) [inline packages/pnpm/src/cmd/outdated.ts]: sortPackages would be the right function name.
Maybe even sortOutdatedPackages

File: packages/pnpm/test/recursive/outdated.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ import {
 const test = promisifyTape(tape)
 const testOnly = promisifyTape(tape.only)
 
-test('pnpm recursive outdated', async (t: tape.Test) => {
+testOnly('pnpm recursive outdated', async (t: tape.Test) => {
```
Comment (zkochan) [inline packages/pnpm/test/recursive/outdated.ts]: testOnly => test

File: packages/pnpm/test/recursive/outdated.ts [hunk 5]
```diff
@@ -51,15 +51,16 @@ test('pnpm recursive outdated', async (t: tape.Test) => {
...
-    └───────────────────┴─────────┴────────┴──────────────────────┘
+    ┌───────────────────┬─────────┬────────┬────────────┐
+    │ Package           │ Current │ Latest │ Dependents │
+    ├───────────────────┼─────────┼────────┼────────────┤
+    │ is-negative       │ 1.0.0   │ 2.1.0  │ project-2  │
+    ├───────────────────┼─────────┼────────┼────────────┤
+    │ is-negative (dev) │ 1.0.0   │ 2.1.0  │ project-3  │
+    ├───────────────────┼─────────┼────────┼────────────┤
+    │ is-positive       │ 1.0.0   │ 3.1.0  │ project-1  │
+    │                   │         │        │ project-3  │
```
Comment (zkochan) [inline packages/pnpm/test/recursive/outdated.ts]: I am not sure how I feel about this change. Maybe it should go through a separate PR

File: packages/pnpm/test/recursive/outdated.ts [hunk 6]
```diff
@@ -69,15 +70,61 @@ test('pnpm recursive outdated', async (t: tape.Test) => {
...
+    project-2
+
+    is-negative (dev)
+    1.0.0 => 2.1.0
+    project-3
+
+    is-positive
+    1.0.0 => 3.1.0
+    project-1
+    project-3
```
Comment (zkochan) [inline packages/pnpm/test/recursive/outdated.ts]: If there will be many (like 50 dependents) then this will be unreadable.
I think at least an explanation should be here.
Something like
```
Dependent: project-1
```

But probably keeping it in one line will fit better into the screen

```
Dependents: project-1, project-2
```

**Comment by ExE-Boss** [comment]
I’ve opened #2027 to&nbsp;try and&nbsp;clean‑up the&nbsp;<code>pnpm&nbsp;outdated</code> and&nbsp;<code>pnpm&nbsp;recursive&nbsp;outdated</code> implementations as&nbsp;a&nbsp;pre‑requisite to&nbsp;this.

**Comment by nikoladev** [comment]
Looks great!

The only things I'd change are:
- making the names bold so it's easier to visually distinguish which package it is about
- making the arrow grey so that it doesn't add visual noise when reading the version numbers

Like so:

image

File: packages/pnpm/src/cmd/outdated.ts (hunk 1 above)
Comment (ExE-Boss) [reply packages/pnpm/src/cmd/outdated.ts:104]: I think that might be because it defaults to `undefined`, and `--no-table` gets parsed as `--table=false`.

Maybe the option should be named `--list`.

File: packages/pnpm/src/cmd/outdated.ts [hunk 7]
```diff
@@ -88,42 +91,72 @@ export default async function (
...
+
+    return table([
+      columnNames,
+      ...sortedPackages(outdatedPackages).map((outdatedPkg) => columnFns.map((fn) => fn(outdatedPkg))),
+    ], TABLE_OPTIONS)
+  } else {
+    return sortedPackages(outdatedPackages)
+      .map((outdatedPkg) => {
+        let info = stripIndent`
+          ${renderPackageName(outdatedPkg)}
```
Comment (nikoladev) [inline packages/pnpm/src/cmd/outdated.ts]: Make package name bold so it's easier to visually distinguish which package it is about:

```suggestion
          ${chalk.bold(renderPackageName(outdatedPkg))}
```

File: packages/pnpm/src/cmd/outdated.ts [hunk 8]
```diff
@@ -88,42 +91,72 @@ export default async function (
...
+          ${renderCurrent(outdatedPkg)} => ${renderLatest(outdatedPkg)}`
```
Comment (nikoladev) [inline packages/pnpm/src/cmd/outdated.ts]: Make arrow grey so that it doesn't add visual noise when reading the version numbers:

```suggestion
          ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
```

File: packages/pnpm/src/cmd/recursive/outdated.ts [hunk 9]
```diff
@@ -89,51 +91,86 @@ export default async (
...
         },
       },
-    }),
+    })
+  } else {
+    output = sortedPackages(outdatedByNameAndType)
+      .map((outdatedPkg) => {
+        let info = stripIndent`
+          ${renderPackageName(outdatedPkg)}
+          ${renderCurrent(outdatedPkg)} => ${renderLatest(outdatedPkg)}`
```
Comment (nikoladev) [inline packages/pnpm/src/cmd/recursive/outdated.ts]: Make arrow grey so that it doesn't add visual noise when reading the version numbers:

```suggestion
          ${renderCurrent(outdatedPkg)} ${chalk.grey('=>')} ${renderLatest(outdatedPkg)}`
```

File: packages/pnpm/src/cmd/recursive/outdated.ts [hunk 10]
```diff
@@ -89,51 +91,86 @@ export default async (
...
+          width: getCellWidth(data, 3, 30)
         },
       },
-    }),
+    })
+  } else {
+    output = sortedPackages(outdatedByNameAndType)
+      .map((outdatedPkg) => {
+        let info = stripIndent`
+          ${renderPackageName(outdatedPkg)}
```
Comment (nikoladev) [inline packages/pnpm/src/cmd/recursive/outdated.ts]: Make package name bold so it's easier to visually distinguish which package it is about:

```suggestion
          ${chalk.bold(renderPackageName(outdatedPkg))}
```

File: packages/pnpm/src/cmd/outdated.ts (hunk 1 above)
Comment (nikoladev) [reply packages/pnpm/src/cmd/outdated.ts:104]: @ExE-Boss I was actually also working on an implementation of this (didn't check if anybody else was working on it 🤦‍♀️) and I also went with `--list` for this reason.

Though I do feel that `--no-table` is much clearer. Also, there is already a `pnpm list` command, so also having `pnpm outdated --list` might confuse people.

**Suggestion**: Maybe we could default `opts.table` to `true` unless `--no-table` is explicitly passed?

File: packages/pnpm/src/cmd/outdated.ts (hunk 3 above)
Comment (aparajita) [reply packages/pnpm/src/cmd/outdated.ts]: No, because **sort**Something implies sorting in place. This is returning a sorted copy, which **sorted** implies. Also, it's intended to be used like this:

```js
for (const pkg of sortedPackages(outdatedPackages))
```

which is more semantic than

```js
for (const pkg of sortPackages(outdatedPackages))
```

File: packages/pnpm/src/cmd/outdated.ts (hunk 1 above)
Comment (aparajita) [reply packages/pnpm/src/cmd/outdated.ts:104]: @ExE-Boss You might as well finish up this feature, I don't have a lot more time to spend on it right now.

File: packages/pnpm/test/recursive/outdated.ts (hunk 6 above)
Comment (aparajita) [reply packages/pnpm/test/recursive/outdated.ts]: Actually, 50 comma-delimited dependents stretched out across multiple lines is completely unreadable. I tried both ways before settling on one per line.

But I would advocate against using a label for dependents only. It will look weird unless all of the items have labels, and at that point you might as well just do `--json`.

Another option would be to put a divider above and below the dependents.

**Comment by nikoladev** [comment]
I've been using this new feature for a bit and I have to say it has improved my package updating experience by a lot!

Thanks!

**Comment by zkochan** [comment]
next: interactive updates
